lib/resty/http.lua

```diff
@@ -914,5 +957,99 @@ function _M.proxy_response(self, response, chunksize)
     until not chunk
 end
 
+function _M.set_proxy_options(self, opts)
+    self.proxy_opts = opts
+end
+
+function _M.get_proxy_uri(self, scheme, host)
+    if not self.proxy_opts then
+        return nil
+    end
+
+    -- Check if the no_proxy option matches this host. Implementation adapted
+    -- from lua-http library (https://github.com/daurnimator/lua-http)
+    if self.proxy_opts.no_proxy then
+        if self.proxy_opts.no_proxy == "*" then
+            -- all hosts are excluded
+            return nil
+        end
+
+        local no_proxy_set = {}
+        -- wget allows domains in no_proxy list to be prefixed by "."
+        -- e.g. no_proxy=.mit.edu
+        for host_suffix in self.proxy_opts.no_proxy:gmatch("%.?([^,]+)") do
```

Comment:
In ngx_lua we're assured access to the `ngx.re.*` regex functions which can be JIT compiled. At the risk of premature optimisation (I haven't tested and compared) I'm pretty sure `gmatch` will be quite a bit slower here. Please consider changing these to use `ngx.re.gmatch`.

Reply:
Rewrote the logic to use ngx.re functions instead of built-in gmatch. I had to make some changes to the logic here as apparently the built-in gmatch and ngx.re.gmatch use completely different regex syntax. Please check carefully that I did not make any mistakes in the process (at leasts tests are passing for me).